### PR TITLE
Feature/459 more label colors

### LINKED ISSUE #459: More label colors
Increase the number of selectable label colors. Colors help to get an overview quickly. 
Option: allow definition of own colors 

Solution accepted by @bruderol, @mi-we, @adiherzog:
- Simply add a text field behind the date picker, where the selected color value can be changed in more details by changing the value in format "#0099ff" (for advanced users for those rare situations, where you need another color).
- Only highlight currently selected color in the picker, if it is available in the picker
- There should be also a preview of an example of the label in currently selected style (color and name).

### PR DESCRIPTION
**Pull request by lynparbone**
Allow the user to configure custom colors by providing a hexadecimal value.

Fixes #459

### CODE REVIEW

File: scenarioo-client/app/manage/label-colors/color-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import * as contrast from 'contrast/index';
+
+export const hexadecimalColorIsValid = (color: string): boolean => {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/color-helpers.ts]: convention is more like isBlablaValid

File: scenarioo-client/app/manage/label-colors/color-helpers.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    // Color hexadecimal should have the format #000 or #0000000
+    const hexadecimalColorRegexPattern = /^#([0-9A-F]{3}){1,2}$/i;
+    return hexadecimalColorRegexPattern.test(color);
+};
+
+export const getContrastingForegroundColorOfBackgroundColor = (backgroundColor: string): string => {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/color-helpers.ts]: does not matter that it's a background color for this function. I would go with getContrastingColor(hexColor: string)

File: scenarioo-client/test/e2e/pages/labelConfigurationsPage.ts [hunk 3]
```diff
@@ -43,6 +43,19 @@ class LabelConfigurationsPage {
         return this.saveButton.click();
     }
 
+    async updateLabelConfigurationWithColorAsHexadecimal(rowIndex, labelName, colorAsHexadecimal) {
```
Comment (danielsuter) [inline scenarioo-client/test/e2e/pages/labelConfigurationsPage.ts]: add types

File: scenarioo-client/app/manage/label-colors/color-helpers.ts (hunk 2 above)
Comment (lynparbone) [reply scenarioo-client/app/manage/label-colors/color-helpers.ts]: Done

File: scenarioo-client/app/manage/label-colors/color-helpers.ts (hunk 1 above)
Comment (lynparbone) [reply scenarioo-client/app/manage/label-colors/color-helpers.ts]: Changed to isHexColorValid, thanks!

File: scenarioo-client/test/e2e/pages/labelConfigurationsPage.ts (hunk 3 above)
Comment (lynparbone) [reply scenarioo-client/test/e2e/pages/labelConfigurationsPage.ts]: Done

**Comment by lynparbone** [comment]
I noticed there is some duplicate code in multiple components regarding the labels.
I've created a separate issue for this here: https://github.com/scenarioo/scenarioo/issues/1016

**Comment by bruderol** [comment]
@lynparbone thanks! Perfect to open issues for things that have to be improved later!

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts [hunk 4]
```diff
@@ -15,16 +15,15 @@ useCase('Configure label colors')
 
         scenario('Create, edit and delete label configurations')
             .it(async () => {
-
                 // Given number of preconfigured colors (in config.xml for demo)
-                const numberOfPreconfiguredColors = 3;
+                const numberOfPreconfiguredColors: number = 3;
```
Comment (bruderol) [inline scenarioo-client/test/e2e/specs/configure_label_colors.ts]: type not needed here  - typescript recognises the type when assigning a value

File: scenarioo-client/app/manage/label-colors/color-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    // Convert e.g. #ABC to #AABBCC
+    return '#' + hexColorValue
+        .split('')
+        .reduce((extendedHexColor: string[], currentHexValue: string) => {
+            return extendedHexColor.concat([currentHexValue, currentHexValue]);
+        }, [])
+        .join('');
+};
```
Comment (bruderol) [inline scenarioo-client/app/manage/label-colors/color-helpers.ts]: Würde daraus eine Klasse mit statischen Methoden machen und sie Color nennen - Color.isValid(...), Color.getRandom(), und Color.getContrasting(..) sieht hübsch aus und lässt sich einfacher importieren als wenn ich die genauen Methodennamen erst raussuchen und einzeln importieren muss.

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts [hunk 6]
```diff
@@ -46,4 +45,25 @@ useCase('Configure label colors')
 
             });
 
+        scenario('Create a label configuration with a custom color')
+            .it(async () => {
+                const customLabelColor: string = '#00FF00';
+                const numberOfPreconfiguredColors: number = 3;
```
Comment (bruderol) [inline scenarioo-client/test/e2e/specs/configure_label_colors.ts]: auch hier sind die Types unnötig - Type inference

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts [hunk 7]
```diff
@@ -46,4 +45,25 @@ useCase('Configure label colors')
...
+
+                await HomePage.goToPage();
+                await step('navigate away from the label config page to some other page');
+
+                await labelConfigurationsPage.navigateToPage();
+                await labelConfigurationsPage.assertNumConfigurations(numberOfPreconfiguredColors + 1);
+                await labelConfigurationsPage.assertConfigurationColor(3, customLabelColor);
+                await step('go back to label config page, label is still there and has the custom color');
+
+            });
```
Comment (bruderol) [inline scenarioo-client/test/e2e/specs/configure_label_colors.ts:70]: random color button wurde nicht geprüft

File: scenarioo-client/app/manage/label-colors/color-helpers.ts (hunk 5 above)
Comment (bruderol) [reply scenarioo-client/app/manage/label-colors/color-helpers.ts]: diese Utility klasse würde ich in ein Utility package legen, falls wir noch mehr mit Farben machen

**Review by bruderol — COMMENTED** [review]
Thanks! Great work! See my minor comments for how to slightly improve. Then it is perfect for me.

Maybe avoid that the last empty line for adding a new label can not be removed ??

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts (hunk 4 above)
Comment (lynparbone) [reply scenarioo-client/test/e2e/specs/configure_label_colors.ts]: Thanks for the info, will keep that in mind

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts (hunk 6 above)
Comment (lynparbone) [reply scenarioo-client/test/e2e/specs/configure_label_colors.ts]: Thanks for the info.

File: scenarioo-client/test/e2e/specs/configure_label_colors.ts (hunk 7 above)
Comment (lynparbone) [reply scenarioo-client/test/e2e/specs/configure_label_colors.ts:70]: Added a check for the random button in the e2e tests.

File: scenarioo-client/app/manage/label-colors/color-helpers.ts (hunk 5 above)
Comment (lynparbone) [reply scenarioo-client/app/manage/label-colors/color-helpers.ts]: This makes sense and does simplify the import. Thank you for pointing it out.
I've created a Color class with static functions and placed it in the shared/utils folder.

**Comment by lynparbone** [comment]
Made the delete button conditional: It is now never displayed on the last line, so it can't be deleted by accident (Note: As soon as the user starts typing a label name, it is displayed on that line as a new line is added)

@bruderol: IMO it is ok to merge now

**Comment by bruderol** [comment]
Thanks! :+1:

**Pull request merged**